apps/snitch_core/test/domain/shipment_engine_test.exs

```diff
@@ -0,0 +1,91 @@
+  use ExUnit.Case, async: true
+  use Snitch.DataCase
+
+  import Snitch.Factory
+  import Snitch.{OrderCase, StockCase, ShipmentEngineCase}
+  alias Snitch.Data.Schema.{Order, StockLocation}
+  alias Snitch.Domain.ShipmentEngine
+
+  @sl_manifest %{
+    "default" => %{
+      default: true
+    },
+    "backup" => %{},
+    "origin" => %{}
+  }
+
+  setup :variants
+  setup :stock_locations
+
+  describe "order fulfilled" do
+    @tag variant_count: 4
+    test "with multiple packages", context do
+      package_types = [[1, 0, 1, 0], [1, 1, 1, 1], [1, 0, 0, 0], [0, 1, 0, 1]]
+      %{variants: vs} = context
+      line_items = line_items_with_price(vs, [1, 1, 1, 1])
+      order = %Order{id: 42, line_items: line_items, shipping_address: insert(:address)}
+      packages = packages(line_items, context, package_types)
+      value = ShipmentEngine.run(packages, order)
+      items = fulfilled_items(value)
+
+      Enum.each(line_items, fn li ->
+        assert Enum.member?(items, li)
+      end)
+    end
+
+    @tag variant_count: 4
+    test "with single package", context do
+      package_types = [[1, 1, 1, 1]]
+      %{variants: vs} = context
+      line_items = line_items_with_price(vs, [1, 1, 1, 1])
+      order = %Order{id: 42, line_items: line_items, shipping_address: insert(:address)}
+      packages = packages(line_items, context, package_types)
+      value = ShipmentEngine.run(packages, order)
+      items = fulfilled_items(value)
+
+      Enum.each(line_items, fn li ->
+        assert Enum.member?(items, li)
+      end)
+    end
+  end
+
+  describe "order unfulfilled" do
+    @tag variant_count: 4
+    test "with multiple packages", context do
+      package_types = [[1, 0, 1, 0], [1, 1, 1, 0], [1, 0, 0, 0], [0, 1, 0, 0]]
+      %{variants: vs} = context
+      line_items = line_items_with_price(vs, [1, 1, 1, 1])
+      order = %Order{id: 42, line_items: line_items, shipping_address: insert(:address)}
+      packages = packages(line_items, context, package_types)
+      assert [] = ShipmentEngine.run(packages, order)
+    end
+
+    @tag variant_count: 4
+    test "with single package", context do
+      package_types = [[1, 1, 1, 0]]
+      %{variants: vs} = context
+      line_items = line_items_with_price(vs, [1, 1, 1, 1])
+      order = %Order{id: 42, line_items: line_items, shipping_address: insert(:address)}
+      packages = packages(line_items, context, package_types)
+      assert [] = ShipmentEngine.run(packages, order)
+    end
+  end
+
+  defp stock_locations(context) do
+    locations =
+      context
+      |> Map.get(:stock_location_manifest, @sl_manifest)
+      |> stock_locations_with_manifest()
+
+    {_, stock_locations} = Repo.insert_all(StockLocation, locations, returning: true)
+
+    [stock_locations: stock_locations]
+  end
+
+  defp fulfilled_items(pckgs) do
+    Enum.reduce(pckgs, [], fn %{items: list}, acc ->
+      Enum.reduce(list, acc, fn %{line_item: item}, acc -> [item | acc] end)
```

Comment:
1. Reads much better as an `Enum.map`.
2. It is much cheaper to make 2 MapSets of variant/item ids and assert that they are equal.

Reply:
had thought of this but creating a `MapSet` of items from packages may lead to skipping of those cases in which lineitems may be repeated in two packages.

Reply:
but that is a guarantee given by this module and, shouldn't that be tested?

Reply:
No, this guarantee is given by `Snitch.Domain.Shipment`, the module that _creates_ the list that is _simply being consumed here,_ in `ShipmentEngine`. 
Hence, you can happily use the `MapSet`.